### PR TITLE
Refactor - Simplify program accounts in transaction loading

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
Currently we skip the loading of program accounts (of top level instructions) by filling them with an empty entry and later loading them separately in `load_executable_accounts()`. This not only causes redundant loading logic for the upgradeable loaders programdata special case, but also duplication of all program accounts, even multiple times if a program is used more than once.

#### Summary of Changes
- Adds the deduplicated built-in programs at the very end (behind the account deps).
- Removes `load_executable_accounts()` and uses the normal account loading instead.
- Removes `test_accounts_account_not_found()` as the `ProgramAccountNotFound` case is now covered by `test_program_sbf_invoke_in_same_tx_as_deployment()` (see #29725).
- Adjusts number of total loaded accounts in `test_load_accounts_multiple_loaders()`.

### CODE REVIEW

**Comment by taozhu-chicago** [comment]
Will it be backported to 1.14?

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -267,144 +268,183 @@ impl Accounts {
...
+            error_counters.account_not_found += 1;
+            return Err(TransactionError::AccountNotFound);
+        }
+
         // Appends the account_deps at the end of the accounts,
         // this way they can be accessed in a uniform way.
         // At places where only the accounts are needed,
         // the account_deps are truncated using e.g:
         // accounts.iter().take(message.account_keys.len())
         accounts.append(&mut account_deps);
```
Comment (taozhu-chicago) [inline runtime/src/accounts.rs:389]: Perhaps should remove this line? Looks like `accounts` is expected to not including `account_deps` during iteration of instructions, and Ln 440 appends `account_deps` again.

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -267,144 +268,183 @@ impl Accounts {
...
+                            self.accounts_db.load_with_fixed_root(ancestors, owner_id)
+                        {
+                            account_deps.push((*owner_id, program_account));
+                        } else {
+                            error_counters.account_not_found += 1;
+                            return Err(TransactionError::ProgramAccountNotFound);
+                        }
+                        owner_index
+                    };
+                    account_indices.push(owner_index as IndexOfAccount);
```
Comment (taozhu-chicago) [inline runtime/src/accounts.rs:431]: in previous version, it calls `account_indices.insert(0, _)` to keep an order of `pd_index_n, account_index_n, ..... pd_index_0, account_index_0`. Not familiar if such order is necessary, just want to point out new version will change how account_indices is ordered.

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -267,144 +268,183 @@ impl Accounts {
...
+                if !account_found {
+                    error_counters.account_not_found += 1;
+                    return Err(TransactionError::ProgramAccountNotFound);
+                }
+                if !program_account.executable() {
+                    error_counters.invalid_program_for_execution += 1;
+                    return Err(TransactionError::InvalidProgramForExecution);
+                }
+                let owner_id = program_account.owner();
+                if !native_loader::check_id(owner_id) {
```
Comment (taozhu-chicago) [inline runtime/src/accounts.rs:414]: Again not familiar enough if call-chain's supported, but would expect `while` instead of `if` here

**Review by taozhu-chicago — COMMENTED** [review]
Nice to avoid redundant accounts loading by tracking with `account_found_and_dep_index`. Just few questions that I am not sure about.

File: runtime/src/accounts.rs (hunk 1 above)
Comment (Lichtso) [reply runtime/src/accounts.rs:389]: Ah, I should rename the second `account_deps` to `builtins` or something. Yes both are added to `accounts` eventually, but the linear search only runs over the second one which should be smaller because there are only a few built-in programs.

File: runtime/src/accounts.rs (hunk 2 above)
Comment (Lichtso) [reply runtime/src/accounts.rs:431]: The loader chain was parsed in reverse and `insert(0, _)` did reverse it again so that it came out "forward". The order is correct (and stays the same as before), otherwise all subsequent tests in the program-runtime would fail with "Invalid Program" as they would try to use a custom BPF program as built-in (if the order was swapped).

File: runtime/src/accounts.rs (hunk 3 above)
Comment (Lichtso) [reply runtime/src/accounts.rs:414]: The thing is: CPI of built-in programs is independent of the loader ownership chains here.

For example, [the upgradeable loader (`BPFLoaderUpgradeab1e11111111111111111111111`) does a CPI call to the system instruction processor (`11111111111111111111111111111111`) to create a new account](https://github.com/solana-labs/solana/blob/2c347ac0a5891ecd2137af9ef08b16a550080369/programs/bpf_loader/src/lib.rs#L637). The upgradeable loader however, is owned by the phony native loader (`NativeLoader1111111111111111111111111111111`) which has no entry in the accounts DB and also no owner, thus no relation to the system instruction processor being called.

In other words, I don't understand the purpose of nested ownership for built-ins.

Comment (jackcmay) [reply runtime/src/accounts.rs:415]: Loader chains from top level instructions is lost here, aka no support for non-builti loaders or 3rd party loafers.  Maybe you are bringing up that non-builtin loaders are not supported by cpi?

Comment (Lichtso) [reply runtime/src/accounts.rs:415]: Yes, kind off, we currently have three mechanisms:
- Top-level instructions (with builtin loader ownership chains)
- CPI
- Syscalls

Program runtime v2 is already unifying CPI and syscalls (by modelling syscalls as CPI to built-in programs).
I hope we can get rid of built-in loader ownership chains too and don't see a reason why 3rd party addons / plugins / mods can't just be treated as built-ins as well and simply have a model where all programs can call all other programs using only one mechanism.

Comment (taozhu-chicago) [reply runtime/src/accounts.rs:415]: From what I see during replaying mainnet snapshots, I have not seen a transaction has call chain deeper than 2. Often `BPFLoaderUpgradeab1e` loads one program (and its programdata) accounts. But it doesn't necessary mean removing support of loader call chain will not break existing user programs. 

Is it possible to continue support chain of loader in this refactor PR, so it does not alter function/behavior. Move the change to separate PR instead?

Comment (Lichtso) [reply runtime/src/accounts.rs:415]: Continuing to support it in this is PR might be possible yes. Still, I have not heard any argument why we would keep it, if nobody can show me where it is used. Also, I don't think it is actually possible to construct such a chain in the current program-runtime as one can not change the ownership of a non-empty or executable account.

Comment (taozhu-chicago) [reply runtime/src/accounts.rs:415]: > I don't think it is actually possible to construct such a chain in the current program-runtime as one can not change the ownership of a non-empty or executable account.

If this is the case, then removal of that logic would be good. The simplified code is much cleaner for sure. I don't know enough about it, but I will continue pay attention while I am replaying ledgers. Will post if I find any case of this loader chain is used. Meanwhile, any comments from someone more familiar with it? @jstarry @jackcmay  perhaps?

Comment (jackcmay) [reply runtime/src/accounts.rs:415]: This issue is being discussed in the `virtual-machine` channel on discord.  The current consensus is that we don't need it for move and don't need to retain it here.  If we remove it here, there is nothing in the protocol that I'm aware of that prevents us from adding this functionality back if we decide otherwise later.

**Review by taozhu-chicago — DISMISSED** [review]
lgtm, thanks for discussing the decision of removing chained-loader.

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -267,144 +268,190 @@ impl Accounts {
...
+                    if let Some(account_index) = account_dep_index {
+                        account_indices.insert(0, *account_index);
+                    }
+                    let owner_id = program_account.owner();
+                    if native_loader::check_id(owner_id) {
+                        return Ok(account_indices);
+                    }
+                    program_index = if let Some(owner_index) = accounts
+                        .get(builtins_start_index..)
+                        .ok_or(TransactionError::ProgramAccountNotFound)?
```
Comment (pgarg66) [inline runtime/src/accounts.rs:426]: Wondering if it'll always return `None` (mapping to `TransactionError::ProgramAccountNotFound`) here.
https://github.com/solana-labs/solana/pull/29728/files#diff-a6df749724ef3698bf0b90625702f100670b58c585f91da8725f78a94b04196eR391 sets `builtins_start_index` to `accounts.len()`. There is no new entry pushed to `accounts` since `builtins_start_index` was initialized.

Comment (Lichtso) [reply runtime/src/accounts.rs:426]: There is a few lines below: `accounts.push((*owner_id, program_account));`

Comment (pgarg66) [reply runtime/src/accounts.rs:426]: But, the code flow will not reach there, right? Because of this `.ok_or(TransactionError::ProgramAccountNotFound)?` it'll just skip the code block.
